pandas/core/groupby.py

```diff
@@ -1047,41 +1021,33 @@ def tail(self, n=5):
         2  b  1
 
         """
-        obj = self._selected_obj
-        rng = np.arange(0, -self.grouper._max_groupsize, -1, dtype='int64')
-        in_tail = self._cumcount_array(rng, ascending=False) > -n
-        tail = obj[in_tail]
-        return tail
+        mask = self._cumcount_array(ascending=False) < n
+        return self._selected_obj[mask]
 
-    def _cumcount_array(self, arr=None, ascending=True):
+    def _cumcount_array(self, ascending=True):
         """
-        arr is where cumcount gets its values from
-
         note: this is currently implementing sort=False (though the default is sort=True)
               for groupby in general
         """
-        if arr is None:
-            arr = np.arange(self.grouper._max_groupsize, dtype='int64')
+        ids, _, ngroups = self.grouper.group_info
+        sorter = _get_group_index_sorter(ids, ngroups)
+        ids, count = ids[sorter], len(ids)
 
-        len_index = len(self._selected_obj.index)
-        cumcounts = np.zeros(len_index, dtype=arr.dtype)
-        if not len_index:
-            return cumcounts
+        if count == 0:
+            return np.empty(0, dtype=np.intp)
 
-        indices, values = [], []
-        for v in self.indices.values():
-            indices.append(v)
+        run = np.r_[True, ids[:-1] != ids[1:]]
+        rep = np.diff(np.r_[np.nonzero(run)[0], count])
+        out = (~run).cumsum()
```

Comment:
actually ``np.intp`` is ok. its this:

```
In [1]: np.array([True,False,True]).cumsum()
Out[1]: array([1, 1, 2])

In [2]: np.array([True,False,True]).cumsum().dtype
Out[2]: dtype('int32')
```

must be some weird casting on windows. providing an accumulator dtype seems to work.
```
In [3]: np.array([True,False,True]).cumsum(dtype='int64').dtype
Out[3]: dtype('int64')
```